cext/cutils.c

```diff
@@ -0,0 +1,186 @@
+
+#ifdef _MSC_VER
+
+typedef __int32 int32_t;
+typedef unsigned __int32 uint32_t;
+typedef __int64 int64_t;
+typedef unsigned __int64 uint64_t;
+
+#else
+#include <stdint.h>
+#endif
+
+static int32_t FNV1_OFFSET_BASIS = 0x811c9dc5;
+static int32_t FNV1_PRIME = 0x01000193;
+
+
+PyObject* hashcode(PyObject* self, PyObject *args);
+PyObject* schema_id(PyObject* self, PyObject *args);
+
+PyObject* str_hashcode(PyObject* data);
+int32_t str_hashcode_(PyObject* data, int lower);
+PyObject* b_hashcode(PyObject* data);
+
+static PyMethodDef methods[] = {
+    {"hashcode", (PyCFunction) hashcode, METH_VARARGS, ""},
+    {"schema_id", (PyCFunction) schema_id, METH_VARARGS, ""},
+    {NULL, NULL, 0, NULL}       /* Sentinel */
+};
+
+static struct PyModuleDef moduledef = {
+    PyModuleDef_HEAD_INIT,
+    "_cutils",
+    0,                    /* m_doc */
+    -1,                   /* m_size */
+    methods,              /* m_methods */
+    NULL,                 /* m_slots */
+    NULL,                 /* m_traverse */
+    NULL,                 /* m_clear */
+    NULL,                 /* m_free */
+};
+
+static char* hashcode_input_err = "supported only strings, bytearrays, bytes and memoryview";
+static char* schema_id_input_err = "input argument must be dict or int";
+static char* schema_field_type_err = "schema keys must be strings";
+
+PyMODINIT_FUNC PyInit__cutils(void) {
+	return PyModule_Create(&moduledef);
+}
+
+PyObject* hashcode(PyObject* self, PyObject *args) {
+    PyObject* data;
+
+    if (!PyArg_ParseTuple(args, "O", &data)) {
+        return NULL;
+    }
+
+    if (data == Py_None) {
+        return PyLong_FromLong(0);
+    }
+    else if (PyUnicode_CheckExact(data)) {
+        return str_hashcode(data);
+    }
+    else {
+        return b_hashcode(data);
+    }
+}
+
+PyObject* str_hashcode(PyObject* data) {
+    return PyLong_FromLong(str_hashcode_(data, 0));
+}
+
+int32_t str_hashcode_(PyObject *data, int lower) {
+    int32_t res = 0;
+    Py_ssize_t sz = PyUnicode_GET_LENGTH(data);
+
+    Py_ssize_t i;
+    for (i = 0; i < sz; i++) {
+        Py_UCS4 ch = PyUnicode_READ_CHAR(data, i);
+
+        if (lower) {
+            ch = Py_UNICODE_TOLOWER(ch);
+        }
+
+        res = 31 * res + ch;
+    }
+
+    return res;
+}
+
+PyObject* b_hashcode(PyObject* data) {
+    int32_t res = 1;
+    Py_ssize_t sz; char* buf;
+
+    if (PyBytes_CheckExact(data)) {
+        sz = PyBytes_GET_SIZE(data);
+        buf = PyBytes_AS_STRING(data);
+    }
+    else if (PyByteArray_CheckExact(data)) {
+        sz = PyByteArray_GET_SIZE(data);
+        buf = PyByteArray_AS_STRING(data);
+    }
+    else if (PyMemoryView_Check(data)) {
+        Py_buffer* pyBuf = PyMemoryView_GET_BUFFER(data);
+        sz = pyBuf->len;
+        buf = (char*)pyBuf->buf;
+    }
+    else {
+        PyErr_SetString(PyExc_ValueError, hashcode_input_err);
+        return NULL;
+    }
+
+    Py_ssize_t i;
+    for (i = 0; i < sz; i++) {
+        res = 31 * res + buf[i];
```

Comment:
I believe we should use `signed char` or `unsigned char` here. Currently we use `char` which has implementation-dependent signedness so the code above could result in different hashcode calculation on different platforms.

Reply:
Good point